x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutJobAction.java

```diff
@@ -15,38 +18,85 @@
+    private final SecurityContext securityContext;
+
 
     @Inject
-    public TransportPutJobAction(TransportService transportService, ClusterService clusterService,
+    public TransportPutJobAction(Settings settings, TransportService transportService, ClusterService clusterService,
                                  ThreadPool threadPool, XPackLicenseState licenseState, ActionFilters actionFilters,
                                  IndexNameExpressionResolver indexNameExpressionResolver, JobManager jobManager,
-                                 AnalysisRegistry analysisRegistry) {
+                                 DatafeedManager datafeedManager, AnalysisRegistry analysisRegistry) {
         super(PutJobAction.NAME, transportService, clusterService, threadPool, actionFilters, PutJobAction.Request::new,
             indexNameExpressionResolver, PutJobAction.Response::new, ThreadPool.Names.SAME);
         this.licenseState = licenseState;
         this.jobManager = jobManager;
         this.analysisRegistry = analysisRegistry;
+        this.datafeedManager = datafeedManager;
+        this.securityContext = XPackSettings.SECURITY_ENABLED.get(settings) ?
+            new SecurityContext(settings, threadPool.getThreadContext()) : null;
     }
 
     @Override
     protected void masterOperation(Task task, PutJobAction.Request request, ClusterState state,
                                    ActionListener<PutJobAction.Response> listener) throws Exception {
-        jobManager.putJob(request, analysisRegistry, state, listener);
+        jobManager.putJob(request, analysisRegistry, state, ActionListener.wrap(
+            jobCreated -> {
+                if (jobCreated.getResponse().getDatafeedConfig().isPresent() == false) {
+                    listener.onResponse(jobCreated);
+                    return;
+                }
+                datafeedManager.putDatafeed(
+                    new PutDatafeedAction.Request(jobCreated.getResponse().getDatafeedConfig().get()),
+                    // Use newer state from cluster service as the job creation may have created shared indexes
+                    clusterService.state(),
+                    licenseState,
+                    securityContext,
+                    threadPool,
+                    ActionListener.wrap(
+                        created -> listener.onResponse(jobCreated),
+                        failed -> jobManager.deleteJob(
+                            new DeleteJobAction.Request(request.getJobBuilder().getId()),
+                            state,
+                            ActionListener.wrap(
+                                deleted -> listener.onFailure(failed),
+                                deleteFailed -> {
+                                    logger.warn(
+                                        () -> new ParameterizedMessage(
+                                            "[{}] failed to cleanup job after datafeed creation failure",
```

Comment:
If we fail here, will we have some maintenance activity cleaning those jobs up?
Or will it be a responsibility of the user?

Reply:
> maintenance activity cleaning those jobs up?

There is no way of knowing if the job was initially created with a datafeed or not. I don't think it wise to have a maintenance action to delete any job that doesn't have a datafeed.

The user will have to intervene here and it SHOULD be a rare occurrence.